File: backend/utils/classifier.js
const db = require('../db');

function classifyTransaction(description, vpa) {
  const categories = db.prepare('SELECT * FROM categories ORDER BY name').all();
  const text = `${description || ''} ${vpa || ''}`.toLowerCase();

  for (const cat of categories) {
    if (cat.name === 'Others') continue;
    const keywords = JSON.parse(cat.keywords || '[]');
    for (const kw of keywords) {
      if (kw && text.includes(kw.toLowerCase())) return cat.id;
    }
  }

  const others = db.prepare("SELECT id FROM categories WHERE name = 'Others'").get();
  return others ? others.id : null;
}

module.exports = { classifyTransaction };

File: android/app/src/main/java/com/thejas/upitracker/utils/Classifier.kt
package com.thejas.upitracker.utils

import com.thejas.upitracker.data.model.Category
import org.json.JSONArray

object Classifier {

    fun classify(description: String?, vpa: String?, categories: List<Category>): Long? {
        val text = "${description.orEmpty()} ${vpa.orEmpty()}".lowercase()

        for (cat in categories) {
            if (cat.name == "Others") continue
            val keywords = parseKeywords(cat.keywords)
            if (keywords.any { kw -> kw.isNotBlank() && text.contains(kw.lowercase()) }) {
                return cat.id
            }
        }
        return categories.firstOrNull { it.name == "Others" }?.id
    }

    private fun parseKeywords(json: String): List<String> = try {
        val arr = JSONArray(json)
        (0 until arr.length()).map { arr.getString(it) }
    } catch (_: Exception) { emptyList() }
}
